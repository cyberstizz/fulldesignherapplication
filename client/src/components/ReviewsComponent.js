import React, { useEffect, useState } from 'react';
import { useParams } from 'react-router';
import ReviewComponent from './ReviewComponent';
import './ReviewsComponent.scss';
import axios from 'axios';

const ReviewsComponent = () => {
  // State to hold the list of reviews
  const [reviews, setReviews] = useState([]);
  // State to hold the average rating
  const [averageRating, setAverageRating] = useState(null);
  // Retrieve productType and productId from URL parameters
  const { productType, productId } = useParams();

  useEffect(() => {
    // Function to fetch reviews from the API
    const fetchReviews = async () => {
      try {
        // Set API URL based on the environment
        const apiUrl = process.env.NODE_ENV === 'production'
          ? 'https://designhercustomekreations-c288e9799350.herokuapp.com'
          : 'http://localhost:3001';

        // Get reviews for the specific product
        const response = await axios.get(`${apiUrl}/reviews/${productType}/${productId}`);
        
        // Update state with reviews and average rating
        setReviews(response.data.reviews);
        setAverageRating(response.data.averageRating);
      } catch (error) {
        console.error('Error fetching reviews:', error);
      }
    };

    fetchReviews();
    console.log(reviews);
  }, [productId, productType]); // Dependencies to re-fetch when productId or productType change

  return (
    <div className="reviewsComponent">
      {/* Display average rating or a placeholder if not available */}
      <h2>Average Rating: {averageRating || 'Not rated yet'}</h2>
      {reviews.length > 0 ? (
        // Map through reviews and render ReviewComponent for each
        reviews.map((review, index) => (
          <ReviewComponent
            key={index}
            headline={review.headline}
            rating={review.rating}
            review={review.review}
            userName={review.userName} // Ensure userName is included in your server response
          />
        ))
      ) : (
        // Message if no reviews are present
        <p>No reviews yet.</p>
      )}
    </div>
  );
};

export default ReviewsComponent;
