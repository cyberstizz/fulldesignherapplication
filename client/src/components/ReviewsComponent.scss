.reviewsComponent {
    display: flex;
    flex-direction: column;
    align-items: center;
    padding: 20px;
  
    @media (min-width: 768px) {
      max-width: 80%;
      margin: 0 auto;
    }
  }
  