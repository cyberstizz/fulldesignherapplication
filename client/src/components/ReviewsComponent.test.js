// src/components/ReviewsComponent.test.js
import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import axios from 'axios';
import ReviewsComponent from './ReviewsComponent';

jest.mock('axios');

const renderComponent = () => {
  return render(
    <BrowserRouter>
      <ReviewsComponent />
    </BrowserRouter>
  );
};

describe('ReviewsComponent', () => {
  const mockReviews = [
    {
      headline: 'Great product!',
      rating: 5,
      review: 'I love this product. Highly recommend!',
      userName: 'User1',
    },
    {
      headline: 'Not bad',
      rating: 3,
      review: 'It is okay, could be better.',
      userName: 'User2',
    },
  ];

  beforeEach(() => {
    axios.get.mockResolvedValue({
      data: {
        reviews: mockReviews,
        averageRating: 4,
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders without crashing', () => {
    renderComponent();
  });

  test('displays the average rating correctly', async () => {
    renderComponent();

    await waitFor(() => {
      expect(screen.getByText('Average Rating: 4')).toBeInTheDocument();
    });
  });

  test('displays reviews when they are available', async () => {
    renderComponent();

    await waitFor(() => {
      expect(screen.getByText('Great product!')).toBeInTheDocument();
      expect(screen.getByText('I love this product. Highly recommend!')).toBeInTheDocument();
      expect(screen.getByText('User1')).toBeInTheDocument();

      expect(screen.getByText('Not bad')).toBeInTheDocument();
      expect(screen.getByText('It is okay, could be better.')).toBeInTheDocument();
      expect(screen.getByText('User2')).toBeInTheDocument();
    });
  });

  test('displays a message when there are no reviews', async () => {
    axios.get.mockResolvedValueOnce({
      data: {
        reviews: [],
        averageRating: null,
      },
    });

    renderComponent();

    await waitFor(() => {
      expect(screen.getByText('No reviews yet.')).toBeInTheDocument();
    });
  });

  test('matches snapshot', async () => {
    const { asFragment } = renderComponent();

    await waitFor(() => {
      expect(screen.getByText('Average Rating: 4')).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  test('handles errors gracefully', async () => {
    axios.get.mockRejectedValueOnce(new Error('Error fetching reviews'));

    renderComponent();

    await waitFor(() => {
      expect(screen.getByText('No reviews yet.')).toBeInTheDocument();
    });

    // Optionally, you could check console.error mock if you want to ensure the error is logged
  });
});
