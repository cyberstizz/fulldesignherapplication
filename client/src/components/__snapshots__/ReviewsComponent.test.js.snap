// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewsComponent matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="reviewsComponent"
  >
    <h2>
      Average Rating: 4
    </h2>
    <div
      class="reviewComponent"
    >
      <h3
        class="headline"
      >
        Great product!
      </h3>
      <div
        class="rating"
      >
        Rating: 5
      </div>
      <p
        class="review"
      >
        I love this product. Highly recommend!
      </p>
      <p
        class="userName"
      >
        - User1
      </p>
    </div>
    <div
      class="reviewComponent"
    >
      <h3
        class="headline"
      >
        Not bad
      </h3>
      <div
        class="rating"
      >
        Rating: 3
      </div>
      <p
        class="review"
      >
        It is okay, could be better.
      </p>
      <p
        class="userName"
      >
        - User2
      </p>
    </div>
  </div>
</DocumentFragment>
`;
